### PR TITLE
AEA-1044 Feature/libp2p node tls support

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

libp2p node tls support, replaces #2563

## Fixes

- #1978

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes and CI passes too
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/helpers/pipe.py [hunk 1]
```diff
@@ -595,6 +608,99 @@ async def close(self) -> None:
         await self._sock.close()
 
 
+class TCPSocketChannelClientTLS(TCPSocketChannelClient):
+    """Interprocess communication channel client using tcp sockets with TLS."""
+
+    VERIFICATION_SIGNATURE_WAIT_TIMEOUT = 5.0
+
+    def __init__(  # pylint: disable=unused-argument
```
Comment (DavidMinarsch) [inline aea/helpers/pipe.py:616]: Why pylint disable?

File: tests/test_helpers/test_pipe/test_pipe.py [hunk 2]
```diff
@@ -137,6 +138,42 @@ async def test_connection_refused(self):
...
+
+@pytest.mark.asyncio
+class TestAEAHelperTCPSocketChannelTLS:
+    """Test that TCPSocketChannelTLS work properly"""
+
+    SERVER_PUB_KEY = (
+        "03e09d7febfc3e3ef4c38321fb10c4303956e65e1d03ead77470941e82428f80d2"
+    )
+    SIGNATURE = b"0D\x02 \x7f\xcf\xa7\xe9~\x1d{\xf1\xb6\"\x98\x1d'\xd2}\x13\xb5\x13\x0c,]}7\xf3\xa6G\x958q_\x89\x08\x02 >\x91\x9e!\xb5\x1f/\xe4\x0c\xab\x9ej\xb7~Z8\x0b\x06Et\x9f\x1b.SZ*Q\xb7\x13\x85\xaa-"
+    SESSION_KEY = b"\x04D\x08\xe1\xca\xfc;\x01(\x0f\xb9&\x92\x12\xe6\x0b\x02\xdc\x082\xa3\xff\x05\x1f#\xeaK\xd3!\xf5\xcc\xcf\x86\x98\x17g\x80_\xc7o;\x9e\x86,O\xbd\xa1bO\x06\x92\x85\x94\x9f\xbf}\xc2\xdd[\xe6AI\x9a\x9c\x92"
```
Comment (DavidMinarsch) [inline tests/test_helpers/test_pipe/test_pipe.py:156]: How do we generate all this when the connection is used in an agent?

File: setup.py [hunk 3]
```diff
@@ -58,7 +58,8 @@ def get_all_extras() -> Dict:
     "pyyaml>=4.2b1,<6.0",
     "requests>=2.22.0,<3.0.0",
     "python-dotenv>=0.14.0,<0.18.0",
-    "ecdsa>=0.15,<0.17.0"
+    "ecdsa>=0.15,<0.17.0",
+    "asn1crypto>=1.4.0,<=1.5.0",
```
Comment (DavidMinarsch) [inline setup.py:62]: Why not `>=1.4.0, <=2.0.0`?

**Review by DavidMinarsch — COMMENTED** [review]
Please add an example test of using this in cli context in an aea.

File: tests/test_helpers/test_pipe/test_pipe.py (hunk 2 above)
Comment (solarw) [reply tests/test_helpers/test_pipe/test_pipe.py:156]: server pub key is from connection configuration
signature generated by libp2p node with private key
session pub key is tmp key generated by libp2p node

File: setup.py (hunk 3 above)
Comment (solarw) [reply setup.py:62]: not sure about how this library is going to be updated, i think this version restriction is ok for the moment

File: aea/helpers/pipe.py (hunk 1 above)
Comment (solarw) [reply aea/helpers/pipe.py:616]: good point, it was copy-pasted

File: tests/test_helpers/test_pipe/test_pipe.py (hunk 2 above)
Comment (DavidMinarsch) [reply tests/test_helpers/test_pipe/test_pipe.py:156]: So the signature (I assume using connection private key) needs to be requested in the context of a cert request? What's the signature of?

File: aea/helpers/pipe.py [hunk 4]
```diff
@@ -595,6 +608,96 @@ async def close(self) -> None:
         await self._sock.close()
 
 
+class TCPSocketChannelClientTLS(TCPSocketChannelClient):
```
Comment (DavidMinarsch) [inline aea/helpers/pipe.py:611]: I'd prefer this code (`TCPSocketChannelClientTLS`) to live in the `p2p_libp2p_client`. Then we don't need a new framework level dependency. Instead it will be a package level dependency.

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go [hunk 5]
```diff
@@ -609,18 +619,89 @@ func (dhtPeer *DHTPeer) Close() []error {
 	return status
 }
 
+func generate_x509_cert() (*tls.Certificate, error) {
+	privBtcKey, err := btcec.NewPrivateKey(elliptic.P256())
```
Comment (DavidMinarsch) [inline libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go:623]: Please provide a docstring, in particular it needs to explain that we use a session key and for what etc.

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go [hunk 6]
```diff
@@ -609,18 +619,89 @@ func (dhtPeer *DHTPeer) Close() []error {
...
 		check(err)
 	}
+
+	cert_pub_key := dhtPeer.cert.PrivateKey.(*ecdsa.PrivateKey).Public().(*ecdsa.PublicKey)
+	cert_pub_key_bytes := elliptic.Marshal(cert_pub_key.Curve, cert_pub_key.X, cert_pub_key.Y)
+	signature, err := dhtPeer.key.Sign(cert_pub_key_bytes)
+	if err != nil {
+		lerror(err).Msgf("while generating tls signature")
+		check(err)
+	}
```
Comment (DavidMinarsch) [inline libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go:703]: extract into helper method?

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer_test.go [hunk 7]
```diff
@@ -1814,10 +1815,15 @@ func SetupDelegateClient(
...
 	client.Rx = make(chan *aea.Envelope, 2)
-	client.Conn, err = net.Dial("tcp", host+":"+strconv.FormatInt(int64(port), 10))
+	conf := &tls.Config{
+		InsecureSkipVerify: true,
+	}
+	client.Conn, err = tls.Dial("tcp", host+":"+strconv.FormatInt(int64(port), 10), conf)
 	if err != nil {
 		return nil, nil, err
 	}
+	// Read signature, skip it here
```
Comment (DavidMinarsch) [inline libs/go/libp2p_node/dht/dhtpeer/dhtpeer_test.go:1825]: Why skip?

File: tests/test_helpers/test_pipe/test_pipe.py (hunk 2 above)
Comment (solarw) [reply tests/test_helpers/test_pipe/test_pipe.py:156]: signature is a session public key value signed with peer private key that can be verified by the peer public key(address)

File: aea/helpers/pipe.py (hunk 4 above)
Comment (solarw) [reply aea/helpers/pipe.py:611]: good point

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go (hunk 5 above)
Comment (solarw) [reply libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go:623]: done

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go (hunk 6 above)
Comment (solarw) [reply libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go:703]: done

File: setup.py [hunk 8]
```diff
@@ -58,7 +58,7 @@ def get_all_extras() -> Dict:
     "pyyaml>=4.2b1,<6.0",
     "requests>=2.22.0,<3.0.0",
     "python-dotenv>=0.14.0,<0.18.0",
-    "ecdsa>=0.15,<0.17.0"
+    "ecdsa>=0.15,<0.17.0",
```
Comment (DavidMinarsch) [inline setup.py:61]: Please remove unnecessary change

File: packages/fetchai/connections/p2p_libp2p_client/connection.py [hunk 9]
```diff
@@ -564,3 +574,93 @@ async def _process_messages(self) -> None:
             self._in_queue.put_nowait(envelope)
             if envelope is None:
                 break  # pragma: no cover
+
+
+class TCPSocketChannelClientTLS(TCPSocketChannelClient):
+    """Interprocess communication channel client using tcp sockets with TLS."""
+
+    VERIFICATION_SIGNATURE_WAIT_TIMEOUT = 5.0
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/p2p_libp2p_client/connection.py:582]: This should also be configurable via kwargs, and be the default value

File: packages/fetchai/connections/p2p_libp2p_client/connection.py [hunk 10]
```diff
@@ -564,3 +574,93 @@ async def _process_messages(self) -> None:
...
+        sock = await self._open_tls_connection()
+        session_pub_key = self._get_session_pub_key(sock.writer)
+
+        try:
+            signature = await asyncio.wait_for(
+                sock.read(), timeout=self.VERIFICATION_SIGNATURE_WAIT_TIMEOUT
+            )
+        except asyncio.TimeoutError:  # pragma: nocover
+            raise ValueError(
+                f"Failed to get peer verification record in timeout: {self.VERIFICATION_SIGNATURE_WAIT_TIMEOUT}"
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/p2p_libp2p_client/connection.py:626]: timeout should be configurable and therefore variable

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer_test.go (hunk 7 above)
Comment (solarw) [reply libs/go/libp2p_node/dht/dhtpeer/dhtpeer_test.go:1831]: proper check added

File: packages/fetchai/connections/p2p_libp2p_client/connection.py (hunk 9 above)
Comment (solarw) [reply packages/fetchai/connections/p2p_libp2p_client/connection.py:582]: added, do we want to configure it with connection config? or stay just configurable for TCPSocketChannelClientTLS?

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer_test.go (hunk 7 above)
Comment (DavidMinarsch) [reply libs/go/libp2p_node/dht/dhtpeer/dhtpeer_test.go:1831]: So the comment is no longer correct...please update

File: packages/fetchai/connections/p2p_libp2p_client/connection.py (hunk 9 above)
Comment (DavidMinarsch) [reply packages/fetchai/connections/p2p_libp2p_client/connection.py]: I think we should be able to configure via connection config too!

File: libs/go/libp2p_node/dht/dhtpeer/dhtpeer_test.go (hunk 7 above)
Comment (solarw) [reply libs/go/libp2p_node/dht/dhtpeer/dhtpeer_test.go:1831]: done

**Review by lrahmani — APPROVED** [review]
LGTM!
I like both go and python TLS connections abstractions.
I highly recommend adding unit/integration tests with invalid session key signature to make sure that tls is properly configured.

**Review by DavidMinarsch — APPROVED** [review]
LGTM
